lsp/src/main/kotlin/org/gradle/declarative/lsp/DeclarativeTextDocumentService.kt

```diff
@@ -385,19 +388,24 @@ private fun completionItem(property: DataProperty, resolvedType: DataType) =
 
 private fun computePropertyByValueFactoryCompletions(
     dataClass: DataClass,
-    analysisSchema: AnalysisSchema,
+    schema: AnalysisSchema,
     valueFactoryIndex: ValueFactoryIndex,
 ): List<CompletionItem> {
     return dataClass.properties.flatMap { property ->
-        val resolvedType = SchemaTypeRefContext(analysisSchema).resolveRef(property.valueType)
+        val typeRefContext = SchemaTypeRefContext(schema)
+        val resolvedType = typeRefContext.resolveRef(property.valueType)
         if (resolvedType is DataType.ClassDataType) {
             val factoriesForProperty = valueFactoryIndex.factoriesForProperty(resolvedType.name)
             factoriesForProperty
+                ?.filter { true } // TODO: need to filter out cases where type parameters don't match the properties type
```

Comment:
> need to filter out cases where type parameters don't match the properties type

This should be solvable by `checkIsAssignable`, which is unfortunately not public either. As probably a better alternative, we might want to introduce a utility that both computes the type substitution and checks assignability. This should be a good facade that saves clients from having to invoke the two low-level type operations in the right order.

Also, `factoriesForProperty` might need to include some type information in the keys in addition to just the FQNs. I don't see an obvious approach; it needs some research.

Reply:
Have extracted and merged the two operations.